### ISSUE TITLE
Can I use cublas for implementation of my custom operator?

### ISSUE THREAD

**Issue by songyuc** [NVIDIA/DALI#4511]
Hi, guys,
I am learning C++ for customizing my own operator. And I find `cublas` is an awesome library for performance of some tensors calculations on GPU.
The demo in the [doc](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/custom_operations/custom_operator/create_a_custom_operator.html#:~:text=%23include%20%3Ccuda_runtime_api.h%3E) used `cuda_runtime_api`, so, I want to confirm whether `cublas` is also suitable here.

Your answer and guide will be appreciated!

**Comment by songyuc** [comment]
Hi @JanuszL ,
I have a half-baked idea that Torchscript-C++ might be used to realize a custom operator of label encoding in DALI, as I might export a C++ operator by reusing my PyTorch codes of label encoding.
But I am not quite familiar with the underlying mechanism, so I don't know whether it would work.
I have also heard that Torchscript is based on the technique of JIT, which might be not compatible with DALI.
So I guess this idea seems not that reasonable, and I just want to confirm it.

**Comment by JanuszL** [comment]
Hi @songyuc,

I have not tried this with Torchscript-C++, but we have a similar concept in DALI done for Numba compiled code. You define a function in python and Numba compiles it either to CPU or GPU code that can be executed as a part of the DALI pipeline. See relevant files [here](https://github.com/NVIDIA/DALI/tree/main/dali/operators/numba_function) and [here](https://github.com/NVIDIA/DALI/tree/main/dali/python/nvidia/dali/plugin/numba).

**Comment by songyuc** [comment]
I am sorry for my not catching up with the latest updates of DALI. My thought might still remain at the previous situation as:
- python (both CPU and GPU) - is the best for PoC but not for performance
- use Numba (only CPU for now) - the best compromise between flexibility and speed
- add a native operator (CPU and GPU) - the best regarding performance and flexibility but requires advanced C++ and CUDA knowledge

So, now DALI support customizing GPU operator with `Numba`, am I right?

**Comment by JanuszL** [comment]
> So, now DALI support customizing GPU operator with Numba, am I right?

Sorry to say but the support for Numba GPU is still very rudimentary.

**Comment by JanuszL** [comment]
@songyuc,

As I said, I'm not that much into TorchScript so I don't fully know how it works under the hood. Still, if the python code calls a C++ function at the end it should be possible to make this call from DALI as well. Still, it requires some research and doing something similar to what we did for Numba.

**Comment by songyuc** [comment]
@JanuszL , thanks sincerely for your patient reply.
Alternatively, inspired by your shared talk, I realize there is a good way to reuse my PyTorch codes as export them into an operator in TensorRT, and then called it in a DALI-GPU operator in a similar way in the [doc](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/custom_operations/custom_operator/create_a_custom_operator.html#:~:text=CUDA_CALL(cudaMemcpyAsync(%0A%20%20%20%20%20%20%20%20%20%20%20%20output.raw_mutable_tensor(sample_idx)%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20input.raw_tensor(sample_idx)%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20shape%5Bsample_idx%5D.num_elements()%20*%20input.type_info().size()%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20cudaMemcpyDeviceToDevice%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20ws.stream()))%3B). 
Do you this idea is reasonable?

**Comment by JanuszL** [comment]
Maybe you can use TRITON and [DALI backend](https://github.com/triton-inference-server/dali_backend), and let the TRITON do this as it can use TensorRT backend as well?

**Comment by JanuszL** [comment]
> In fact, I want to use DALI to accelerate my training process, instead of deployment (sorry for not clarifying this).

In this case, TensorRT may be too heavy for this. I would check if you can use DALI arithmetic operators to achieve what you need. If that doesn't work I would try the native code implementation.

**Comment by songyuc** [comment]
@JanuszL , appreciate your suggestion sincerely.
As I need to encode each object instance into a mask target tensor, for-loops is necessary, so maybe C++ operator is best choice in order to achieve a faster speed. 
I would continue learning cublas and try to implement my custom operator of label encoding.
